contracts/IdentityOracle.sol

```diff
@@ -0,0 +1,308 @@
+
+contract IdentityOracle is DAOUpgradeableContract, ChainlinkClient, Ownable {
+
+    address public dao_avatar; // this implementation is only to test. In live it would be replaced for dao.avatar
+
+    bytes32 public stateHash; // current state hash
+    string  public stateDataIPFS; // ipfs cid
+
+    bytes32 public tmpStateHash; // this is used for the first of three steps of oracle updating process
+    bytes32 public tmpStateDataIPFS1; // this is used for the second of three steps of oracle updating process
+    bytes32 public tmpStateDataIPFS2; // this is used for the second of three steps of oracle updating process
+
+    uint256 private constant ORACLE_PAYMENT = 10**17;
+    string  private JOBID_GET_STATE_HASH = "f8336d36386945059b6039d9f407c827";
+    string  private JOBID_GET_IPFS_CID1 = "6fdec1f580574313a73fcd3224528711";
+    string  private JOBID_GET_IPFS_CID2 = "c6910a08a1594b82aa217db00aa0a486";
+    address private constant ORACLE_ADDRESS = 0x7bccD524Fd2Cc264Dd80E4589B4B3409fb61daC0;
+
+    uint256 public lastStartUpdProcInvoked;
+
+    struct WhitelistProofState {
+        bytes32[] proof;
+        uint256 lastProofDate;
+        uint256 lastAuthenticatedDate;
+    }
+
+    mapping(address => WhitelistProofState) private whitelistProofState;
+    address [] private whitelistProofStateAddresses; // It's used to store of the whitelistProofState keys as helper to be able to iterate all items.
+    event ProofResult(bool);
+    mapping(address => bool) private oracleState; // Store oracle address ad if isAllowed
+
+    constructor(address _link) Ownable() {
+        if (_link == address(0)) {
+            setPublicChainlinkToken();
+        } else {
+            setChainlinkToken(_link);
+        }
+        //setChainlinkToken(0xa36085F69e2889c224210F603D836748e7dC0088);  // It's Link address of Kovan
+
+        dao_avatar = msg.sender;
+        oracleState[ORACLE_ADDRESS] = true;
+        oracleState[msg.sender] = true;
+    }
+
+    function _onlyOracle() internal view {
+        require(oracleState[msg.sender], "only allowed oracle can call this method");
+    }
+
+    function _onlyAvatarTmp() internal view {
+        require(
+            address(dao_avatar) == msg.sender,
+            "only avatar can call this method"
+        );
+    }
+
+    //- only the DAO can approve/remove an oracle. onlyAvatar is defined in DAOUpgradeableContract
+    function setOracle(address _oracle, bool _isAllowed) public {
+        _onlyAvatarTmp();
+        oracleState[_oracle] = _isAllowed;
+    }
+
+    function setState(bytes32 _merklehash, string memory _ipfscid)
+        public
+    //- only approved oracles can set the new merkle state plus link to ipfs data used to create state
+    {
+        _onlyOracle();
+        stateHash = _merklehash;
+        stateDataIPFS = _ipfscid;
+    }
+
+    // It's is the first function to be called by the oracle to update IPFS CID and StatetHash
+    function startIPFSandStateHashProcess() public {
```

Comment:
i dont understand why the oracle needs to call 4 different functions. need to keep it to minimum of calls, otherwise it will just be expensive

Reply:
It will be improved using this https://docs.chain.link/docs/large-responses/ in the next Pull Request